engine/engine.go

```diff
@@ -334,7 +349,64 @@ func (e *Engine) sendDHMetadata(ctx context.Context, putMetaReq dhstore.PutMetad
 	return nil
 }
 
+func (e *Engine) sendDHKeyShardMerges(ctx context.Context, merges []dhstore.Merge) error {
+	mergeReq := dhstore.MergeIndexRequest{
+		Merges: make([]dhstore.Merge, 1),
+	}
+	b64Enc := base64.RawStdEncoding
+	var keyBuf []byte
+
+	start := time.Now()
+
+	// Send a separate dhstore.MergeIndexRequest for each merge.
+	for i := range merges {
+		mergeReq.Merges[0] = merges[i]
+
+		// TODO: The Value of all merges is the same. Encode it once and then
+		// Use json.RawMessage to avoid endocing it again.
+		data, err := json.Marshal(mergeReq)
+		if err != nil {
+			return err
+		}
+
+		req, err := http.NewRequestWithContext(ctx, http.MethodPut, e.dhMergeURL, bytes.NewBuffer(data))
+		if err != nil {
+			return err
+		}
+		req.Header.Set("Content-Type", "application/json")
+
+		// Reuse buffer for base64 encoding.
+		key := merges[i].Key
+		encLen := b64Enc.EncodedLen(len(key))
+		if encLen > len(keyBuf) {
+			keyBuf = make([]byte, encLen)
+		}
+		b64Enc.Encode(keyBuf, key)
+		req.Header.Set(shardKeyHeader, string(keyBuf[:encLen]))
```

Comment:
Shall we use base58?

Reply:
Unless the data was already encoded in base58, or needed to be use as base58 somewhere, I would prefer base64.